### ISSUE TITLE
Synchronous document validation for all format

### ISSUE THREAD

**Issue by MarinPostma** [meilisearch/transplant#142]
@Kerollmops and I have come to an agreement on the fact that document formatting must be validated synchronously, so format related errors a returned to the user early. This seems the best choice to do UX wise. This is valid for all document format supported by MeilliSearch, now and in the future (json, csv, jsonl..).

This means that we are effectively validating the documents twice: once before adding them to the document queue, and once more when the update is popped from the queue and actually deserialized.

More than being a performance/memory waste, this means that we are actually doing twice more or less the same thing at two different places in the code base.

We haven't found a trivial design that would make it work nicely yet, but we have decided that I would take some time to think about an elegant solution it is not a priority to expose support for additional document type (not until 0.22 at least). If we can't find a satistying solution, we will resolve to do the validation twice, as it is already the case for JSON [here](https://github.com/meilisearch/transplant/blob/main/meilisearch-http/src/index_controller/update_actor/actor.rs#L132-L143).

@gmourier FYI

**Comment by gmourier** [comment]
Thank you for the information. I will still write the specs for CSV and Jsonl and mention that in the tech part.

**Comment by gmourier** [comment]
Can't it be a quick win to propose new file support for our users? Keeping in mind that doing the validation twice also concerns JSON and can be refactored in a near future?

**Comment by curquiza** [comment]
Closing this, the implementation will be needed in next versions when the spec about NDJSON and CSV will need to be implemented.
